execution-engine/src/fs.rs

```diff
@@ -998,18 +993,27 @@ impl FileSystem {
     /// how a particular execution engine handles the memory.
     /// That is, different engines provide different API to interact the linear memory
     /// space of WASM. Hence, the method here return the read bytes as `Vec<u8>`.
-    pub(crate) fn fd_pread(
+    pub(crate) fn fd_pread<B: AsMut<[u8]>>(
         &self,
         fd: Fd,
-        buffer_len: usize,
+        bufs: &mut [B],
         offset: FileSize,
-    ) -> FileSystemResult<Vec<u8>> {
+    ) -> FileSystemResult<usize> {
         self.check_right(&fd, Rights::FD_READ)?;
         let inode = self.fd_table.get(&fd).ok_or(ErrNo::BadF)?.inode;
 
-        self.lock_inode_table()?
-            .get(&inode)?
-            .read_file(buffer_len, offset)
+        let f = self.lock_inode_table()?;
+        let f = f.get(&inode)?;
+
+        let mut offset = offset;
+        let mut len = 0;
+        for buf in bufs {
+            let delta = f.read_file(buf.as_mut(), offset)?;
+            offset += u64::try_from(delta).unwrap();
```

Comment:
When does this fail?

Reply:
These really can't fail, it's just safer than using `delta as u64` in case someone does something crazy like run Veracruz on a 16-bit system. Or more likely some bug introducing some out-of-range value.

Just like bound-checking, these should (in theory (emphasis on theory)) get elided in many cases by the compiler.

I can adopt `try_from_or_errno` here.

Reply:
Hmm, alternatively could we add a conversion from TryFromIntError for Errno? Then `u64::try_from(x)?` would be valid here.

Reply:
Ah, no we can't since ErrNo is outside of the current crate. Hmm.

I went ahead and changed it to a trait so it's a bit easier to replace `<_>::try_from(x).unwrap()` calls with, but let me know if you want me to revert that. Should be updated now to use `<_>::try_from_or_errno(x)?`